src/core/NewDomain.ml

```diff
@@ -0,0 +1,1898 @@
+
+  let unleash f v =
+    match !v.rel with
+    | Some r -> let d = f r !v.data in v := {data = d; rel = None}; d
+    | None -> !v.data
+
+  let fold f v = f !v.rel !v.data
+end
+
+type 'a face = dim * dim * 'a Lazy.t Delayed.t
+type 'a sys = 'a face list
+type 'a abs = Abs of Name.t * 'a
+
+(** This is the type of "prevalues", that is the domain from which values come.
+    A prevalue can be judged to be a value with respect to a restriction / dimension
+    theory. I think we will not write code that checks that something is a value, but instead
+    make this involved in the pre-and-post- conditions of the semantic operations. *)
+type con =
+  | Pi of quantifier
+  | Sg of quantifier
+  | Ext of ext_clo
+
+  | Lam of clo
+  | Cons of value * value
+  | ExtLam of nclo
+
+  | HCom of {r : dim; r' : dim; ty : hcom_shape; cap : value; sys : con abs sys}
+  | Coe of {r : dim; r' : dim; ty : coe_shape; cap : value}
+  | Com of {r : dim; r' : dim; ty : coe_shape; cap : value; sys : con abs sys}
+
+  | Univ of {kind : Kind.t; lvl : Lvl.t}
+
+  | V of {r : dim; ty0 : value; ty1 : value; equiv : value}
+  | VIn of {r : dim; el0 : value; el1 : value}
+
+  | Neu of {ty : value; neu : neu; sys : con sys}
+
+and value = con Delayed.t
+
+and quantifier =
+  {dom : value;
+   cod : clo}
+
+and coe_shape =
+  [ `Pi of quantifier abs   (** coe in pi *)
+  | `Sg of quantifier abs   (** coe in sigma *)
+  | `Ext of ext_clo abs     (** coe in extension type *)
+  ]
+
+and com_shape = coe_shape
+
+and hcom_shape =
+  [ `Pi of quantifier   (** hcom in pi *)
+  | `Sg of quantifier   (** hcom in sigma *)
+  | `Ext of ext_clo     (** hcom in extension type *)
+  | `Pos                (** fhcom, i.e. hcom in positive types *)
+  ]
+
+
+and head =
+  | Lvl of int
+  | Var of {name : Name.t; twin : Tm.twin; ushift : int}
+  | Meta of {name : Name.t; ushift : int}
+  | NCoe of {r : dim; r' : dim; ty : neu abs; cap : value}
+  | NCom of {r : dim; r' : dim; ty : neu abs; cap : value; sys : con abs sys}
+
+and frame =
+  | FunApp of value
+  | Fst
+  | Snd
+  | ExtApp of dim list
+  | NHCom of {r : dim; r' : dim; cap : value; sys : con abs sys}
+  | VProj of {r : dim; func : value}
+
+and neu =
+  {head : head;
+   frames : frame bwd}
+
+and cell =
+  | Val of con Lazy.t Delayed.t
+  | Dim of dim
+
+and env = {globals : GlobalEnv.t; cells : cell bwd}
+
+and clo = Clo of {bnd : Tm.tm Tm.bnd; env : env}
+and nclo = NClo of {bnd : Tm.tm Tm.nbnd; env : env}
+and ext_clo = ExtClo of {bnd : (Tm.tm * (Tm.tm, Tm.tm) Tm.system) Tm.nbnd; env : env}
+
+
+(** Permutations *)
+module Perm :
+sig
+  type t
+  val freshen_name : Name.t -> Name.t * t
+  val freshen_names : Name.t bwd -> Name.t bwd * t
+  val swap_name : t -> Name.t -> Name.t
+  val fold : (Name.t -> Name.t -> 'a -> 'a) -> t -> 'a -> 'a
+end =
+struct
+  type t = (Name.t * Name.t) list (* favonia: a demonstration of my laziness *)
+  let mimic x = Name.named (Name.name x)
+  let freshen_name x =
```

Comment:
As a matter of style, I like to put newlines between functions, so that in vim I can navigate the code using `{` and `}`.

Reply:
fixed (?)